File: today/pascals-triangle.ts
/**
 * https://leetcode.com/problems/pascals-triangle/
 */
export { generate };

function generate(numRows: number): number[][] {
  if (numRows === 1) return [[1]];

  const result: number[][] = [[1]];
  for (let i = 1; i < numRows; i++) {
    let next: number[] = Array(i + 1);
    result.push(next);
    next[0] = 1;
    next[next.length - 1] = 1;
    for (let j = 1; j < next.length - 1; j++) {
      const lastRow = result[i - 1];
      next[j] = lastRow[j - 1] + lastRow[j];
    }
  }

  return result;
}

File: returned/python/today10/pascals_triangle.py
from typing import List

class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 1:
            return [[1]]
        
        result = [[1],[1,1]]

        for i in range(2, numRows):
            next = [1]
            for j in range(0, i-1):
                next.append(result[i-1][j] + result[i-1][j+1])
            next.append(1)
            result.append(next)
            
        return result